### PR TITLE
feat(FPLA-2355): Enable linking of application to order

### PR DESCRIPTION
**Pull request by jvictorcf**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2355

### Change description ###
Enable linking of application to order.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: ccd-definition/CaseEventToFields/CareSupervision/manageOrders/3a_hearingDetailsSection.json [hunk 1]
```diff
@@ -62,5 +62,44 @@
     "PageFieldDisplayOrder": 4,
     "FieldShowCondition": "orderTempQuestions.hearingDetails=\"YES\" AND manageOrdersApprovedAtHearing=\"Yes\"",
     "ShowSummaryChangeOption": "Y"
+  },
+  {
+    "LiveFrom": "03/06/2021",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "manageOrders",
+    "CaseFieldID": "manageOrdersShouldAllowLinkingApplication",
```
Comment (samuelzerozero) [inline ccd-definition/CaseEventToFields/CareSupervision/manageOrders/3a_hearingDetailsSection.json:70]: without having to declare an extra yes/no, how about implement a logic in the prepopulator similar to CloseCaseBlockPrePopulator 

   return Map.of("orderTempQuestions", orderTempQuestions.toBuilder().closeCase("NO").build());

File: ccd-definition/CaseField/CareSupervision/manageOrders/manageOrdersFields.json [hunk 2]
```diff
@@ -46,6 +46,33 @@
     "SecurityClassification": "Public",
     "Searchable": "N"
   },
+  {
+    "LiveFrom": "03/06/2021",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "manageOrdersShouldAllowLinkingApplication",
```
Comment (samuelzerozero) [inline ccd-definition/CaseField/CareSupervision/manageOrders/manageOrdersFields.json:52]: same, not needed.

File: ccd-definition/AuthorisationCaseField/CareSupervision/judiciary.json [hunk 3]
```diff
@@ -167,6 +167,27 @@
     "UserRole": "caseworker-publiclaw-judiciary",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "03/06/2021",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
```
Comment (samuelzerozero) [inline ccd-definition/AuthorisationCaseField/CareSupervision/judiciary.json:172]: same, not needed.

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json [hunk 4]
```diff
@@ -167,6 +167,27 @@
     "UserRole": "caseworker-publiclaw-gatekeeper",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "03/06/2021",
```
Comment (samuelzerozero) [inline ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json:171]: same, not needed.

File: ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json [hunk 5]
```diff
@@ -181,6 +181,27 @@
     "UserRole": "caseworker-publiclaw-courtadmin",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "03/06/2021",
```
Comment (samuelzerozero) [inline ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json:185]: use     "LiveFrom": "01/01/2017",

File: ccd-definition/ComplexTypes/CareSupervision/OrderTempQuestions.json [hunk 6]
```diff
@@ -151,5 +151,14 @@
     "ElementLabel": " ",
     "FieldShowCondition": "cafcassJurisdictions=\"DO_NOT_SHOW\"",
     "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "03/06/2021",
+    "ID": "OrderTempQuestions",
+    "ListElementCode": "linkApplicationQuestionBlock",
```
Comment (samuelzerozero) [inline ccd-definition/ComplexTypes/CareSupervision/OrderTempQuestions.json:158]: better to drop the suffix "QuestionBlock"? something like "linkApplication" ?

File: ccd-definition/ComplexTypes/CareSupervision/GeneratedOrder.json [hunk 7]
```diff
@@ -168,5 +168,13 @@
     "FieldType": "TextArea",
     "ElementLabel": "Reason for removal",
     "SecurityClassification": "Public"
+  },
+  {
```
Comment (samuelzerozero) [inline ccd-definition/ComplexTypes/CareSupervision/GeneratedOrder.json:172]: should this be visible?

File: e2e/tests/adminManagesOrders_test.js [hunk 8]
```diff
@@ -258,14 +268,18 @@ Scenario('Interim supervision order (C35B)', async ({I, caseViewPage, manageOrde
...
+  //Add additional applications
+  await manageDocumentsForLAHelper.uploadC2(I, caseViewPage, uploadAdditionalApplicationsEventPage);
   await caseViewPage.goToNewActions(config.administrationActions.manageOrders);
 
   await manageOrdersEventPage.selectOperation(manageOrdersEventPage.operations.options.create);
   await I.goToNextPage();
   await manageOrdersEventPage.selectOrder(manageOrdersEventPage.orders.options.c47a);
   await I.goToNextPage();
   manageOrdersEventPage.selectRelatedToHearing(manageOrdersEventPage.hearingDetails.linkedToHearing.options.no);
+  const linkedApplicationId = await manageOrdersEventPage.linkApplication();
```
Comment (samuelzerozero) [inline e2e/tests/adminManagesOrders_test.js:281]: does it need to be visible? I don't think we can assert that

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/dynamic/DynamicListElement.java [hunk 9]
```diff
@@ -16,6 +17,7 @@
 @Data
 @Jacksonized
 @Builder
+@EqualsAndHashCode
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/dynamic/DynamicListElement.java:20]: i think @Data already give you this?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulator.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        return OrderQuestionBlock.LINK_APPLICATION;
+    }
+
+    @Override
+    public Map<String, Object> prePopulate(CaseData caseData) {
+        Map<String, Object> prePopulatedFields = new HashMap<>();
+
+        DynamicList applicationsToLink = caseData.buildApplicationBundlesDynamicList();
+        if (!applicationsToLink.getListItems().isEmpty()) {
+            prePopulatedFields.put("manageOrdersShouldAllowLinkingApplication", "Yes");
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulator.java:25]: case use the same trick as per closeCasePrePopulator without introducing a new casefield

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryServiceTest.java [hunk 11]
```diff
@@ -237,11 +244,11 @@ void generateWithPreviousOrdersWithSameApprovalDate() {
         }
 
         @Test
-        void generateWithPreviousLegacyOrdersWithoutApprovalDate() {
+        void generateWithPreviousLegacyOrdersWithoutApprovalDate_WithLinkedApplication() {
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryServiceTest.java:247]: missing the test when the link is empty?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            .applicationType(OtherApplicationType.C1_PARENTAL_RESPONSIBILITY)
+            .build();
+        caseDataBuilder.additionalApplicationsBundle(singletonList(element(
+            AdditionalApplicationsBundle.builder()
+                .c2DocumentBundle(c2DocumentBundle)
+                .otherApplicationsBundle(otherApplicationsBundle)
+                .build()
+        )));
+
+        Map<String, Object> prePopulatedData = classUnderTest.prePopulate(caseDataBuilder.build());
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java:53]: I'd personally use "actual" rather than prePopulatedData for visual ease of tests

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        assertThat(prePopulatedData).hasSize(2)
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java:55]: can use a stronger assert like isEqualsTo()?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    LinkApplicationBlockPrePopulator classUnderTest = new LinkApplicationBlockPrePopulator();
+
+    @BeforeEach
+    void setUp() {
+        caseDataBuilder = CaseData.builder();
+    }
+
+    @Test
+    void shouldReturnRightQuestionBlock() {
+        AssertionsForClassTypes.assertThat(classUnderTest.accept()).isEqualTo(OrderQuestionBlock.LINK_APPLICATION);
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java:34]: can drop AssertionsForClassTypes

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/dynamic/DynamicListElement.java (hunk 9 above)
Comment (jvictorcf) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/dynamic/DynamicListElement.java:20]: You're right. I'll remove it.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryServiceTest.java (hunk 11 above)
Comment (jvictorcf) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryServiceTest.java:247]: All the other scenarios do not have a linked application.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java (hunk 13 above)
Comment (jvictorcf) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulatorTest.java:55]: as per our conversation, I'll leave this one if that's ok (given it's a complex type we're asserting).

File: e2e/tests/adminManagesOrders_test.js [hunk 15]
```diff
@@ -258,14 +268,18 @@ Scenario('Interim supervision order (C35B)', async ({I, caseViewPage, manageOrde
   });
 });
 
-Scenario('Create C47A appointment of a Children\'s Guardian', async ({I, caseViewPage, manageOrdersEventPage}) => {
+Scenario('Create C47A appointment of a Children\'s Guardian', async ({I, caseViewPage, manageOrdersEventPage, uploadAdditionalApplicationsEventPage}) => {
+  //Add additional applications
+  await manageDocumentsForLAHelper.uploadC2(I, caseViewPage, uploadAdditionalApplicationsEventPage);
```
Comment (samuelzerozero) [inline e2e/tests/adminManagesOrders_test.js:273]: can maybe use the prepared data rather than using the ui to add the c2? (speeding up)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulator.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        return OrderQuestionBlock.LINK_APPLICATION;
+    }
+
+    @Override
+    public Map<String, Object> prePopulate(CaseData caseData) {
+        Map<String, Object> prePopulatedFields = new HashMap<>();
+
+        DynamicList applicationsToLink = caseData.buildApplicationBundlesDynamicList();
+        if (!applicationsToLink.getListItems().isEmpty()) {
+            prePopulatedFields.put("linkApplication", "YES");
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/LinkApplicationBlockPrePopulator.java:25]: linkApplication needs to be at nested level

**Review by samuelzerozero — DISMISSED** [review]

File: e2e/tests/adminManagesOrders_test.js (hunk 15 above)
Comment (jvictorcf) [reply e2e/tests/adminManagesOrders_test.js:273]: I will create a tech debt ticket for this one.

**Review by samuelzerozero — DISMISSED** [review]
Looks good

**Review by samuelzerozero — APPROVED** [review]

File: e2e/pages/events/manageOrders.page.js [hunk 17]
```diff
@@ -283,16 +283,29 @@ const selectCafcassRegion = region => {
   I.click(`${section4.cafcassRegion.group}-${region}`);
 };
 
-const selectEnglandOffice= office => {
+const selectEnglandOffice = office => {
   I.selectOption(section4.englandOffices, office);
 };
 
+const linkApplication = async (applicationId) => {
```
Comment (samuelzerozero) [inline e2e/pages/events/manageOrders.page.js:290]: minor: as an e2e test, is maybe more readeable chosing the option via value rather than key?

**Review by samuelzerozero — APPROVED** [review]

File: e2e/pages/events/manageOrders.page.js (hunk 17 above)
Comment (jvictorcf) [reply e2e/pages/events/manageOrders.page.js]: Good shout. Done.

**Review by samuelzerozero — DISMISSED** [review]

**Review by samuelzerozero — APPROVED** [review]

File: e2e/pages/events/manageOrders.page.js [hunk 18]
```diff
@@ -283,16 +283,28 @@ const selectCafcassRegion = region => {
   I.click(`${section4.cafcassRegion.group}-${region}`);
 };
 
-const selectEnglandOffice= office => {
+const selectEnglandOffice = office => {
   I.selectOption(section4.englandOffices, office);
 };
 
+const linkApplication = async (applicationToChoose) => {
```
[the remark this answers is not in the document]
Comment (jvictorcf) [reply e2e/pages/events/manageOrders.page.js:290]: You're right. Not anymore.
I'll change it.

File: e2e/pages/events/manageOrders.page.js [hunk 19]
```diff
@@ -283,16 +283,28 @@ const selectCafcassRegion = region => {
...
+  I.see('Is there an application for the order on the system?');
+  I.dontSee('Applications');
+  I.checkOption('Yes', '#manageOrdersShouldLinkApplication');
+  I.see('Applications');
```
[the remark this answers is not in the document]
Comment (jvictorcf) [reply e2e/pages/events/manageOrders.page.js:294]: Fair enough. I'll make the change.

[the remark this answers is not in the document]
Comment (jvictorcf) [reply e2e/pages/events/manageOrders.page.js:294]: Actually, @Rob-Doherty. Thinking about it, this element shows up as soon as the element above it is clicked (i.e. no page is reloaded). Do you still think it's worth having a "waitForText()" in this case?

[the remark this answers is not in the document]
Comment (Rob-Doherty) [reply e2e/pages/events/manageOrders.page.js:294]: I was thinking that even without the network calls it could just render slower since it'll be running in a VM for the nightly browsers, but there's a built-in 200ms delay in `stepListener.js` helper that I'd forgotten about so lets go with your original line and we can always change it later if it's actually needed

File: e2e/pages/events/manageOrders.page.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (jvictorcf) [reply e2e/pages/events/manageOrders.page.js]: Done.

**Review by SooryaSu — DISMISSED** [review]

**Review by Rob-Doherty — APPROVED** [review]
tested and looks to work as expected. PO approved

**Review by samuelzerozero — APPROVED** [review]
